Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.cpp

```diff
@@ -0,0 +1,191 @@
+    if (!backObject) {
+        WTFLogAlways("Failed to create GBM buffer of size %dx%d: %s", size.width(), size.height(), safeStrerror(errno).data());
+        return;
+    }
+    UnixFileDescriptor backFD(gbm_bo_get_fd(backObject), UnixFileDescriptor::Adopt);
+    Vector<EGLAttrib> attributes = {
+        EGL_WIDTH, gbm_bo_get_width(backObject),
+        EGL_HEIGHT, gbm_bo_get_height(backObject),
+        EGL_LINUX_DRM_FOURCC_EXT, gbm_bo_get_format(backObject),
+        EGL_DMA_BUF_PLANE0_FD_EXT, backFD.value(),
+        EGL_DMA_BUF_PLANE0_OFFSET_EXT, gbm_bo_get_offset(backObject, 0),
+        EGL_DMA_BUF_PLANE0_PITCH_EXT, gbm_bo_get_stride(backObject),
+        EGL_NONE
+    };
+    m_backImage = display.createEGLImage(EGL_NO_CONTEXT, EGL_LINUX_DMA_BUF_EXT, nullptr, attributes);
+    if (!m_backImage) {
+        WTFLogAlways("Failed to create EGL image for DMABuf with file descriptor: %d", backFD.value());
+        gbm_bo_destroy(backObject);
+        return;
+    }
+    auto* frontObject = gbm_bo_create(WebCore::GBMDevice::singleton().device(), size.width(), size.height(), uint32_t(DMABufFormat::FourCC::ARGB8888), 0);
+    if (!frontObject) {
+        WTFLogAlways("Failed to create GBM buffer of size %dx%d: %s", size.width(), size.height(), safeStrerror(errno).data());
+        display.destroyEGLImage(m_backImage);
+        m_backImage = nullptr;
+        gbm_bo_destroy(backObject);
+        return;
+    }
+    UnixFileDescriptor frontFD(gbm_bo_get_fd(frontObject), UnixFileDescriptor::Adopt);
+    attributes[7] = frontFD.value();
+    m_frontImage = display.createEGLImage(EGL_NO_CONTEXT, EGL_LINUX_DMA_BUF_EXT, nullptr, attributes);
+    if (!m_frontImage) {
+        WTFLogAlways("Failed to create EGL image for DMABuf with file descriptor: %d", frontFD.value());
+        gbm_bo_destroy(frontObject);
+        display.destroyEGLImage(m_backImage);
+        m_backImage = nullptr;
+        gbm_bo_destroy(backObject);
+        return;
+    }
+
+    m_webPage.send(Messages::AcceleratedBackingStoreDMABuf::Configure(WTFMove(backFD), WTFMove(frontFD), gbm_bo_get_format(backObject), gbm_bo_get_offset(backObject, 0), gbm_bo_get_stride(backObject), size));
+    gbm_bo_destroy(backObject);
+    gbm_bo_destroy(frontObject);
+}
+
+void AcceleratedSurfaceDMABuf::willRenderFrame()
+{
+    if (!m_backImage)
+        return;
+
+    glBindTexture(GL_TEXTURE_2D, m_texture);
+    glEGLImageTargetTexture2DOES(GL_TEXTURE_2D, m_backImage);
+    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, m_texture, 0);
+    glBindFramebuffer(GL_FRAMEBUFFER, m_fbo);
+}
+
+void AcceleratedSurfaceDMABuf::didRenderFrame()
+{
+    if (!m_backImage)
+        return;
+
+    glFlush();
+    m_webPage.sendWithAsyncReply(Messages::AcceleratedBackingStoreDMABuf::Frame(), [this, weakThis = WeakPtr { *this }, runLoop = Ref { RunLoop::current() }]() mutable {
```

Comment:
What exactly is problematic? I thought it was thread safe.

Reply:
WebPage isn't usable across different threads. Currently these calls go through the IPC::MessageSender interface and the underlying Connection retrieval and the message-sending act end up working fine. But instead of relying on that, it would be easier on the mind if you retain a Connection ref and use that.

WebSWContextManagerConnection does that. It also runs a WorkQueue on which it receives the IPC messages, which is something we'd want here, but the composition thread is run through a RunLoop. So if desired, in the future a sideline WorkQueue would be receiving the messages and dispatching them on the composition thread.
https://github.com/WebKit/WebKit/blob/main/Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp#L118